refactor(compiler): remove remaining usage of getMutableClone

crisbeto opened the pull request:
Uses an alternate approach of preserving default imports that doesn't involve the `getMutableClone` function that is being removed in TypeScript 5.0.

The alternate approach was already used in the downlevel transform and it works by patching the EmitResolver of the current transformation context to tell TypeScript to preserve the import.

alfaproject commented on packages/compiler-cli/src/ngtsc/imports/src/default.ts:
```diff
@@ -57,29 +59,31 @@ export function getDefaultImportDeclaration(expr: WrappedNodeExpr<unknown>): ts.
   /**
-   * A `Map` which tracks the `Set` of `ts.ImportDeclaration`s for default imports that were used in
-   * a given `ts.SourceFile` and need to be preserved.
+   * A `Map` which tracks the `Set` of `ts.ImportClaus`s for default imports that were used in
```
```suggestion
   * A `Map` which tracks the `Set` of `ts.ImportClause`s for default imports that were used in
```

crisbeto commented:
[Passing TGP](http://test/OCL:509793102:BASE:509820873:1676473043463:8cddb809).

alxhub commented on packages/compiler-cli/src/ngtsc/imports/src/default.ts:
```diff
@@ -89,65 +93,25 @@ export class DefaultImportTracker {
-    // Save memory - there's no need to keep these around once the transform has run for the given
-    // file.
-    this.sourceFileToUsedImports.delete(originalSf);
+        if (clausesForFile) {
```
In the compiler code, explicitly check for `undefined`
```suggestion
        if (clausesForFile !== undefined) {
```

crisbeto committed:
refactor(compiler): remove remaining usage of getMutableClone

Uses an alternate approach of preserving default imports that doesn't involve the `getMutableClone` function that is being removed in TypeScript 5.0.

The alternate approach was already used in the downlevel transform and it works by patching the EmitResolver of the current transformation context to tell TypeScript to preserve the import.

crisbeto commented:
Caretaker note: this passed TGP earlier today and since then there have been only minor changes that won't affect the functionality.

alxhub commented:
This PR was merged into the repository by commit f3f139942cf80704f7a70e3bc5b7ab94566736c5.
